d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,583 @@
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class XdsClientImpl extends XdsClient
+{
+  private static final Logger _log = LoggerFactory.getLogger(XdsClientImpl.class);
+
+  private final Map<String, ResourceSubscriber> _d2NodeSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> _d2NodeMapSubscribers = new HashMap<>();
+
+  private final Node _node;
+  private final ManagedChannel _managedChannel;
+  private final ScheduledExecutorService _executorService;
+  private final BackoffPolicy.Provider _backoffPolicyProvider = new ExponentialBackoffPolicy.Provider();
+
+  private BackoffPolicy _retryBackoffPolicy;
+  private AdsStream _adsStream;
+  private boolean _shutdown;
+
+  public XdsClientImpl(Node node, ManagedChannel managedChannel, ScheduledExecutorService executorService)
+  {
+    _node = node;
+    _managedChannel = managedChannel;
+    _executorService = executorService;
+  }
+
+  @Override
+  void watchXdsResource(String resourceName, ResourceType type, ResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      Map<String, ResourceSubscriber> resourceSubscriberMap = getResourceSubscriberMap(type);
+      ResourceSubscriber subscriber = resourceSubscriberMap.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe {} resource {}", type, resourceName);
+        subscriber = new ResourceSubscriber(type, resourceName);
+        resourceSubscriberMap.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(type, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  public void startRpcStream()
+  {
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(_managedChannel);
+    _adsStream = new AdsStream(stub);
+    _adsStream.start();
+    _log.info("ADS stream started");
+  }
+
+  @Override
+  void shutdown()
+  {
+    _executorService.execute(() ->
+    {
+      _shutdown = true;
+      _log.info("Shutting down");
+      if (_adsStream != null) {
+        _adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+      }
+    });
+
+    _executorService.shutdown();
```

Comment:
Nit: it may be better to let the owner to shutdown this executor. E.g. maybe let XdsLB class to own the thread, and let it close the executor when it is shutdown.